crates/core_arch/src/x86/sse2.rs

```diff
@@ -1461,60 +1461,21 @@ pub unsafe fn _mm_movemask_epi8(a: __m128i) -> i32 {
 #[inline]
 #[target_feature(enable = "sse2")]
 #[cfg_attr(test, assert_instr(pshufd, imm8 = 9))]
-#[rustc_args_required_const(1)]
+#[rustc_legacy_const_generics(1)]
 #[stable(feature = "simd_x86", since = "1.27.0")]
-pub unsafe fn _mm_shuffle_epi32(a: __m128i, imm8: i32) -> __m128i {
-    // simd_shuffleX requires that its selector parameter be made up of
-    // constant values, but we can't enforce that here. In spirit, we need
-    // to write a `match` on all possible values of a byte, and for each value,
-    // hard-code the correct `simd_shuffleX` call using only constants. We
-    // then hope for LLVM to do the rest.
-    //
-    // Of course, that's... awful. So we try to use macros to do it for us.
-    let imm8 = (imm8 & 0xFF) as u8;
+pub unsafe fn _mm_shuffle_epi32<const imm8: i32>(a: __m128i) -> __m128i {
```

Comment:
Is this a breaking change?

Reply:
Yep, same thing happened in a previous PR. We're _kinda_ allowed to file it as "bugfix" though, because the api shouldn't accept a value outside 0..=255

Reply:
yes, similarly to `_mm_shuffle_ps` discussed here https://github.com/rust-lang/stdarch/pull/1018#discussion_r584005413 which was acceptable as a bug fix if I understand correctly ?

Reply:
Well, we could make it still accept any value and do `&0xFF` if we wanted to preserve the bug.

Reply:
sure, whichever you all prefer. my understanding is that a crater run will ultimately help decide whether the bug should be preserved if need be ?

Reply:
Yeah. I think it's fine for now, my comment was more to BurntSushi in case they were worried about breaks in practice.

Reply:
(interestingly this time the immediate is not an `unsigned int` but an `int` in [the intrinsics guide](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_shuffle_epi32&expand=5144) 🤔 )

Reply:
yeah it's just wrong sometimes, or it's right and the headers were specified wrong to begin with.

Reply:
But this isn't just about what the immediate value is right? Doesn't this break all uses since it changes the number of parameters to the function?

Reply:
The old form is rewritten to const generics, see rust-lang/rust#82447 for details.

Reply:
Oh, that, I believe that the new attribute it's getting will let rustc magically rewrite calls using the old style to be the new style instead.

Reply:
Though because of the limits of const generics, some old uses might be broken by this rewriting.

there's a thread about this on Zulip in the consy eval stream. though I don't have the link at hand on mobile, I'll ping you.

Reply:
Ah I see. If this is only going to break things where imm is not a constant, then I'm okay with that and agree it is a bug fix. I am only concerned about function params being changed.

I guess i would also be concerned with how this renders in rustdoc, but whatever the case, that seems fixable.

Reply:
Some initial discussion about this feature can be also be seen in [this other zulip link](https://rust-lang.zulipchat.com/#narrow/stream/268952-edition-2021/topic/Replacing.20rustc_args_required_const.20with.20const.20generics), leading to the rustc PR linked above.

Reply:
Note that `#[rustc_args_required_const]` already required the value to be a constant.

Reply:
Does that mean this change cannot break anything?

Reply:
There is one specific situation where it can break: you can't use an expression derived from generic arguments in a const item (e.g. `const` or a const generic parameter). But `#[rustc_args_required_const]` accepted this.

Example
```rust
fn foo<const X: i32>(a: __m128) {
	// This was previously accepted but will now fail to compile.
	stdarch::_mm_shuffle_ps(a, a, X + 1);

	// This still works.
	stdarch::_mm_shuffle_ps(a, a, X);
}
```